scripts/Stretch.py

```diff
@@ -0,0 +1,357 @@
+                    current_e = self.getValue(line,"E",current_e)
+                    current_f = self.getValue(line,"F",current_f)
+                    onestep = GCodeStep(1,current_x,current_y,current_z,current_e,current_f,comment)
+                elif self.getValue(line,"G") == 92:
+                    current_x = self.getValue(line,"X",current_x)
+                    current_y = self.getValue(line,"Y",current_y)
+                    current_z = self.getValue(line,"Z",current_z)
+                    current_e = self.getValue(line,"E",current_e)
+                    current_f = self.getValue(line,"F",current_f)
+                    onestep = GCodeStep(-1,current_x,current_y,current_z,current_e,current_f,comment)
+                else:
+                    onestep = GCodeStep(-1,current_x,current_y,current_z,current_e,current_f,line);
+                if current_z != layer_z:
+                    Logger.log("d","Layer Z " + "{:.3f}".format(layer_z) + " " + str(len(layerSteps)) + " steps")
+                    if len(layerSteps):
+                        retdata.append(self.processLayer(layerSteps))
+                    layerSteps = []
+                    layer_z = current_z
+                layerSteps.append(onestep)
+        if len(layerSteps):
+            retdata.append(self.processLayer(layerSteps))
+        return retdata
+
+    def processLayer(self,layerSteps):
+        layergcode = ""
+        self.vd1 = np.empty((0,2)) #Start points of segments of already deposited material for this layer
+        self.vd2 = np.empty((0,2)) #End points of segments of already deposited material for this layer
+        current_e = layerSteps[0].e
+        vPos = np.empty((0,2))
+        iflush = 0
+        for i in range(len(layerSteps)):
+            if i==0:
+                current_e = layerSteps[i].e
+            if current_e == layerSteps[i].e:
+                vTrans = np.copy(vPos)
+                if len(vPos) >= 2:
+                    self.workOnSequence(vPos,vTrans)
+                layergcode = self.generate(layerSteps,iflush,i,vTrans,layergcode)
+                iflush = i
+                vPos = np.empty((0,2))
+            if layerSteps[i].step == 0 or layerSteps[i].step == 1:
+                vPos = np.concatenate([vPos,np.array([[layerSteps[i].x,layerSteps[i].y]])])
+            current_e = layerSteps[i].e
+        if len(vPos):
+            vTrans = np.copy(vPos)
+        if len(vPos)>=2:
+            self.workOnSequence(vPos,vTrans)
+        layergcode = self.generate(layerSteps,iflush,len(layerSteps),vTrans,layergcode)
+        return layergcode
+
+    def dumpPos(self,onestep):
+        sout = ""
+        if onestep.f != self.output_f:
+            self.output_f = onestep.f
+            sout += " F"
+            sout += "{:.0f}".format(self.output_f).rstrip(".")
+        if onestep.x != self.output_x or onestep.y != self.output_y or onestep.z != self.output_z:
+            assert onestep.x >= 0 and onestep.x < 200 # Security
```

Comment:
It also doesn't work for printers which use `machine_center_is_zero`, because those may give negative coordinates ;)